cilium/CFP-32810-hybrid-routing-mode.md

```diff
@@ -0,0 +1,83 @@
+# CFP-32810: Add hybrid routing mode in Cilium
+
+**SIG: SIG-NAME** SIG-Clustermesh
+
+**Begin Design Discussion:** 2025-06-16
+
+**Cilium Release:** N/A
+
+**Authors:** Anubhab Majumdar <anmajumdar@microsoft.com>, Vamsi Kalapala <vakr@microsoft.com>, Krunal Jain <krunaljain@microsoft.com>
+
+## Summary
+
+Currently, Cilium supports two modes routing - [encapsulation](https://docs.cilium.io/en/stable/network/concepts/routing/#encapsulation) and [native-routing](https://docs.cilium.io/en/stable/network/concepts/routing/#native-routing). When workloads are on cross subnets with no direct connection, Cilium needs to run in encapsulation mode. This is not efficient, as every packet needs to be encapsulated, even when they are on same subnet, thus paying the MTU overhead and reducing network throughput. This CFP proposes adding a third option to allow Cilium dataplane to make the decision at runtime on whether to route the packet natively or encapsulate.
+
+## Motivation
+
+Consider any of the following scenarios:
+
+1. Two Kubernetes cluster connected through cluster mesh. The nodes are backed by different subnets in two different VNETs, which are connected through VNET peering. The pods are assigned IP address from overlay network.
+2. In a large Kubernetes cluster, pod addresses are assigned from two different subnets with no direct connection.
+
+In either case, Cilium needs to run in encapsulation routing mode. However, when routing packets in above clusters, in many situations, we can get by native routing, and not pay the MTU overhead that comes with encapsulation. The reason for this inflexibility comes from the fact that Cilium datapath is not aware of the underlying network topology, and hence cannot make a decision between the two routing mode during runtime.
+
+With higher adoption in cluster mesh and large scale clusters being common to support AI workloads, it would be beneficial to have a more performant dataplane that can optimize the throughput, while providing support for both encapsulation and native-routing.
+
+## Goals
+
+* Retain the behavior of current routing modes as-is
+* Allow users to set a new routing mode called `hybrid`
+* Allow users to pass in network topologies expressed through CIDR ranges during start-up
+
+## Non-Goals
+
+* Allow dynamic update to CIDR ranges (update without restart)
+
+## Proposal
+
+### Overview
+
+Add a new option for routing-mode config as follows `routing-mode: hybrid`. Also, add a new field in config to pass information about the CIDRs and connectivity. The CIDR information will persist in pinned eBPF maps so that routing decision can be made in runtime by eBPF programs.
```

Comment:
Currently we have `routing-mode: native` + native routing CIDR that determines the directly reachable subnets (which implies that Cilium then masquerades to other destinations). I wonder if there's some conceptual overlap.

Reply:
Right, the idea is kind-of similar. In hybrid mode, we would allow native routing to ranges configured via option. Otherwise encapsulate/pass to Linux stack.

Reply:
Similar or same?

I understand the idea that if we add another flag with its own parameters, that makes it easiest to avoid breaking some other existing functionality. At the same time, if the native routing CIDR determines the directly reachable subnet (and hence excludes traffic from SNAT/masquerade) and then we force the user to also separately configure the `subnet-topology`, then it seems like we're just asking the user to provide the same information twice in two different flags just because of some internal implementation details rather than because of the logical concept that's being configured.

Reply:
I think this will be answered in other threads below from my latest review, so I'll resolve this thread.